api-client/APIClient.test.ts

```diff
@@ -0,0 +1,81 @@
+import { z } from "zod"
+import { createApiClient } from "./api-client"
+import { GenericApiClient } from "./api-endpoint-types"
+
+const ApiSchema = {
+  "/users/{userId}": {
+    GET: {
+      pathParamsType: z.object({ userId: z.string() }),
+      queryParamsType: z.object({ includePosts: z.number().int() }),
+      responseType: z.object({ user: z.string() })
+    }
+  },
+  "/posts": {
+    POST: {
+      payloadType: z.object({ title: z.string(), content: z.string() }),
+      responseType: z.object({ postId: z.string() })
+    }
+  }
+}
+
+type ApiClient = GenericApiClient<typeof ApiSchema>
+
+describe("createApiClient", () => {
+  const originalFetch = global.fetch
+
+  beforeEach(() => {
+    global.fetch = jest.fn()
+  })
+
+  afterEach(() => {
+    jest.restoreAllMocks()
+    global.fetch = originalFetch
+  })
+
+  it("calls the API with the correct method, endpoint, headers, and path params", async () => {
+    const apiClient: ApiClient = createApiClient(
+      ApiSchema,
+      "https://example.com",
+      global.fetch
+    )
+
+    try {
```

Comment:
Had trouble using a mock response from the fetch, it appears as undefined

Reply:
You could use something like [msw](https://github.com/mswjs/msw) which is a pretty convenient way to mock network requests without having to mock fetch.